File: pair-with-target-sum.go

import ()
type Solution struct {
}
// search - finds a pair in arr with a given targetSum.
func (sol *Solution) search(arr []int, targetSum int) []int {
    i, j := 0, len(arr) - 1
    for i < j {
      if sum := arr[i] + arr[j]; sum > targetSum {
        j--
      } else if sum < targetSum {
        i++
      } else {
        return []int{i, j}
      }
    }
    return []int{-1, -1}
}

File: pair-with-target-sum.cc
using namespace std;

#include <iostream>
#include <vector>

class Solution {

  public:

  static vector<int> search(const vector<int> &arr, int targetSum) {
    int i = 0;
    int j = arr.size() - 1;

    while (i < j && arr[i] + arr[j] != targetSum) {
      if (arr[i] + arr[j] < targetSum) {
        ++i;
      } else {
        --j;
      }
    }

    if (i < j) return {i, j};
    else return {-1, -1};
  }

};
